### PR TITLE
bus/ISA: Added HP82321A BASIC Language coprocessor

### PR DESCRIPTION
**Pull request by svenschnelle**
This adds support for the HP82321A BLP cards sold by HP about 35 years ago. Note that this needs a quirk in the new68k to extend a 68k bus cycle until the x86 host served the request. I'm happy for smarter ways to achive this.

### CODE REVIEW

File: src/devices/bus/isa/hpblp.cpp [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+		return 0xb0;
+	return 0xd0;
+}
+
+uint16_t isa8_hpblp_device::bus_r(offs_t offset, uint16_t mem_mask)
+{
+	offset <<= 1;
+
+	if (machine().side_effects_disabled())
+		return 0xffff;
```
Comment (cuavas) [inline src/devices/bus/isa/hpblp.cpp:158]: Would it be better to return `m_bus_data` if it isn’t an address that triggers a bus error so you at least see something in the debugger?

File: src/devices/bus/isa/hpblp.cpp [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+	m_maincpu->defer_access();
+	return 0xffff;
+}
+
+void isa8_hpblp_device::bus_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	offset <<= 1;
+
+	if (machine().side_effects_disabled())
+		return;
```
Comment (cuavas) [inline src/devices/bus/isa/hpblp.cpp:186]: Do you really want to do this?  Unless there’s a really good reason, it’s usually fine to allow writes to work from the debugger.

**Review by cuavas — COMMENTED** [review]
I don’t think `defer_access` that’s added to devcpu is the right thing to do – you want to end the timeslice as soon as possible and allow other CPUs to catch up, but you don’t necessarily want to burn all remaining cycles in the timeslice as your bus acknowledge could come before then depending on how quickly the host responds.  @galibert how can this be done cleanly?

File: src/devices/bus/isa/hpblp.h [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	void reg5_w(offs_t offset, uint8_t data);
+	void reg6_w(offs_t offset, uint8_t data);
+	void reg7_w(address_space &space, offs_t offset, uint8_t data);
+
+	uint8_t status_val(offs_t offset) const;
+
+	DECLARE_WRITE_LINE_MEMBER(hpblp_interrupt);
+	TIMER_CALLBACK_MEMBER(timer10ms);
+	emu_timer *m_timer_10ms;
+	int m_irq;
```
[the remark this answers is not in the document]
Comment (svenschnelle) [reply src/devices/bus/isa/hpblp.h:61]: I'm assuming you're talking about these members, as all the other are already explicitly sized types:

bool m_installed;                                                                                                                                                                     
bool m_ack_buscycle;                                                                                                                                                                  
bool m_bus_read;
bool m_bus_write;
bool m_irq_state;
bool m_reset;

So this would mean i need to use uint8_t or similar here? Or maybe i'm just missing some other members?

[the remark this answers is not in the document]
Comment (cuavas) [reply src/devices/bus/isa/hpblp.h]: The only one that’s a problem is `m_irq`.  Although C++ `bool` isn’t explicitly sized in the C++ standard, it’s defined as being 1 byte in the Itanium C++ ABI (unlike the C99 Boolean which is 4 bytes on PowerPC and various other targets).  Even if we find an ABI where `bool` isn’t one byte, it can be packed/unpacked in on save/load anyway.

TL;DR just make `m_irq` a `uint8_t` and it’ll be fine.

**Comment by cuavas** [comment]
Can you please not force-push each time you update the branch?  It makes it hard to see what’s changed on each update, so the entire changeset has to be reviewed again on each update if you do that.

**Comment by galibert** [comment]
I don't see any way short of defer_access and a small-enough quantum.  The alternative would be a special mode in the scheduler which would essentially be "halted until a callback says to restart" to test at every scheduling.  I haven't dared going into the scheduler yet though.  I know I'll have to, but one nightmare at a time.

**Comment by rb6502** [comment]
@svenschnelle Hi, this has been merged, could you please submit the ROMs?   I couldn't find them on hpmuseum.
